crates/uv/tests/it/version.rs

```diff
@@ -0,0 +1,1137 @@
+
+    let pyproject_toml = context.temp_dir.child("pyproject.toml");
+    pyproject_toml.write_str(
+        r#"
+[project]
+name = "myproject"
+version = "1.10.31"
+requires-python = ">=3.12"
+"#,
+    )?;
+
+    uv_snapshot!(context.filters(), context.version()
+        .arg("minor"), @r"
+    success: false
+    exit_code: 2
+    ----- stdout -----
+
+    ----- stderr -----
+    error: expected version to start with a number, but no leading ASCII digits were found
+    ");
+
+    let pyproject = fs_err::read_to_string(&pyproject_toml)?;
+    assert_snapshot!(
+        pyproject,
+    @r#"
+    [project]
+    name = "myproject"
+    version = "1.10.31"
+    requires-python = ">=3.12"
+    "#
+    );
+    Ok(())
+}
+
+// Dynamic version should error on read
+#[test]
+fn version_get_dynamic() -> Result<()> {
+    let context = TestContext::new("3.12");
+
+    let pyproject_toml = context.temp_dir.child("pyproject.toml");
+    pyproject_toml.write_str(
+        r#"
+        [project]
+        name = "myproject"
+        dynamic = ["version"]
+        requires-python = ">=3.12"
+        "#,
+    )?;
+
+    uv_snapshot!(context.filters(), context.version(), @r"
+    success: false
+    exit_code: 2
+    ----- stdout -----
+
+    ----- stderr -----
+    error: There is no 'project.version' field in: pyproject.toml
+    ");
+
+    let pyproject = fs_err::read_to_string(&pyproject_toml)?;
+    assert_snapshot!(
+        pyproject,
+    @r#"
+    [project]
+    name = "myproject"
+    dynamic = ["version"]
+    requires-python = ">=3.12"
+    "#
+    );
+    Ok(())
+}
+
+// Dynamic version should error on write
+#[test]
+fn version_set_dynamic() -> Result<()> {
+    let context = TestContext::new("3.12");
+
+    let pyproject_toml = context.temp_dir.child("pyproject.toml");
+    pyproject_toml.write_str(
+        r#"
+        [project]
+        name = "myproject"
+        dynamic = ["version"]
+        requires-python = ">=3.12"
+        "#,
+    )?;
+
+    uv_snapshot!(context.filters(), context.version()
+        .arg("0.1.2"), @r"
+    success: false
+    exit_code: 2
+    ----- stdout -----
+
+    ----- stderr -----
+    error: There is no 'project.version' field in: pyproject.toml
+    ");
+
+    let pyproject = fs_err::read_to_string(&pyproject_toml)?;
+    assert_snapshot!(
+        pyproject,
+    @r#"
+    [project]
+    name = "myproject"
+    dynamic = ["version"]
+    requires-python = ">=3.12"
+    "#
+    );
+    Ok(())
+}
+
+// Should fallback to `uv --version` if this pyproject.toml isn't usable for whatever reason
+// (In this case, because tool.uv.managed = false)
+#[test]
+fn version_get_fallback() -> Result<()> {
+    let context = TestContext::new("3.12");
+
+    let pyproject_toml = context.temp_dir.child("pyproject.toml");
+    pyproject_toml.write_str(
+        r#"
+        [project]
+        name = "myapp"
+        version = "0.1.2"
+        
+        [tool.uv]
+        managed = false
```

Comment:
Can we test this where there's not a project as well? The `managed = false` case is pretty niche

Reply:
I wasn't sure if "project missing" was something we could reliably enforce (given we spider into parent dirs)

Reply:
Hmm... I guess during testing we should stop discovery at the uv test directory (`UV_INTERNAL__TEST_DIR`)?

I'm sure we have tests that count on not finding a `pyproject.toml` in a parent though.